File: cpp/hap_cpp2/src/pair_verify.cpp
#include "pair_verify.h"
#include "pairing_helper.h"
#include <stdlib.h>
#include <memory.h>
#include "tlv.h"

//M2
HttpReply PairVerify::stage1(const buf& ios_Curve25519_pk)
{
//        System.out.println("Starting pair verification for " + cfg.getName());
//        this.ios_Curve25519_pk = ios_Curve25519_pk;
//
//        //1. Generate new, random Curve25519 key pair.
//        byte[] acc_Curve25519_sk = new byte[32];
//        secureRandom.nextBytes(acc_Curve25519_sk);

    uint8_t _acc_Curve25519_pk[32];
    for (size_t i = 0; i < sizeof(_acc_Curve25519_pk); ++i)
    {
        _acc_Curve25519_pk[i] = 0x32;
    }
    buf acc_Curve25519_pk;
    acc_Curve25519_pk.data = _acc_Curve25519_pk;
    acc_Curve25519_pk.length = sizeof (_acc_Curve25519_pk);


//        Curve25519.keygen(acc_Curve25519_pk, null, acc_Curve25519_sk);
//
//        //2. Generate the shared secret, SharedSecret, from the acc's Curve25519 secret key and the iOS device's Curve25519  public key
//        sharedSecret = new byte[32];
//        Curve25519.curve(sharedSecret, acc_Curve25519_sk, ios_Curve25519_pk);
//
//        //3. Construct AccessoryInfo = Accessory's Curve25519 PK + AccessoryPairingID + iOS device's Curve25519 public key
//        byte[] accessoryInfo = PairSetup.joinBytes(acc_Curve25519_pk, cfg.getAccPairingID().getBytes(), ios_Curve25519_pk);
//
//        //4. Use Ed25519 to generate AccessorySignature by signing AccessoryInfo with AccessoryLTSK
//        byte[] acc_sign = new EdsaSigner(cfg.getAccPrivateKey()).sign(accessoryInfo);

    uint8_t _acc_sign[64];
    for (size_t i = 0; i < sizeof(_acc_sign); ++i)
    {
        _acc_sign[i] = 0x64;
    }

    buf acc_sign;
    acc_sign.data = _acc_sign;
    acc_sign.length = sizeof (_acc_sign);

    printf("verify1: acc proof length: %d\n", acc_sign.length);

    //5. Construct a sub-TLV
    const char* acc_pairing_ID = "111-11-111";
    buf plaintext = PairingHelper::verify1_encode(acc_pairing_ID, strlen(acc_pairing_ID), acc_sign);

    //6. Derive the symmetric session encryption key, sessionKey, from the Curve25519 shared secret
//        HKDFBytesGenerator hkdf = new HKDFBytesGenerator(new SHA512Digest());
//        hkdf.init(new HKDFParameters(sharedSecret, "Pair-Verify-Encrypt-Salt".getBytes(), "Pair-Verify-Encrypt-Info".getBytes()));
//        sessionKey = new byte[32];
//        hkdf.generateBytes(sessionKey, 0, 32);
//
//        //7. Encrypt the sub-TLV, encryptedData, and generate the 16-byte auth tag, authTag
//        ChachaEncoder chacha = new ChachaEncoder(sessionKey, "PV-Msg02".getBytes());
//        byte[] ciphertext = chacha.encodeCiphertext(plaintext);

    uint8_t _ciphertext [plaintext.length + 16];
    memcpy (&_ciphertext, plaintext.data, plaintext.length);

    for (int i = 0; i < 16; ++i)
    {
        _ciphertext[plaintext.length + i] = i;
    }

    buf ciphertext;
    ciphertext.data = _ciphertext;
    ciphertext.length = sizeof (_ciphertext);

    printf("verify1: encrypted data length: %d\n", ciphertext.length);

    return PairingHelper::verify1_reply(ciphertext, acc_Curve25519_pk);
}

    //M4
HttpReply PairVerify::stage2(const buf& message, const buf& authData)
{
    //2. Decrypt the sub-TLV in encryptedData

    //        ChachaDecoder chacha = new ChachaDecoder(sessionKey, "PV-Msg03".getBytes());
    //        byte[] plaintext = chacha.decodeCiphertext(authData, message);

    buf plaintext = message;

    TLV_Decoder d;
    d.decode(plaintext);
    buf ios_pairing_ID = d.getData(ID);
    buf ios_Ed25519_proof = d.getData(SIGNATURE);

    //        //3. Look up ios PK by ios pairing ID
    //        byte[] ios_Ed25519_ltpk = cfg.getIosPublicKey(new String(ios_pairing_ID));
    //        if (ios_Ed25519_ltpk == null)
    //        {
    //            System.out.println("Unknown user: " + new String(ios_pairing_ID));
    //            //throw new Exception ("unknown user");
    //            return HttpReply.generateAuthError(4);
    //        }
    //
    //        byte[] iosDeviceInfo = PairSetup.joinBytes(ios_Curve25519_pk, ios_pairing_ID, acc_Curve25519_pk);
    //
    //        if (new EdsaVerifier(ios_Ed25519_ltpk).verify(iosDeviceInfo, ios_Ed25519_proof))
    {
        //          cfg.storeReadKey(createKey("Control-Write-Encryption-Key"));
        //          cfg.storeWriteKey(createKey("Control-Read-Encryption-Key"));

        printf("Completed pair verification\n");
        return PairingHelper::verify2_reply();
    }
    //        else
    //        {
    //            System.out.println("Invalid signature. Could not pair " + cfg.getName());
    //            return HttpReply.generateAuthError(4);
    //        }
}

//    private byte[] createKey(String keyType)
//    {
//        HKDFBytesGenerator hkdf = new HKDFBytesGenerator(new SHA512Digest());
//        hkdf.init(new HKDFParameters(sharedSecret, "Control-Salt".getBytes(), keyType.getBytes()));
//        byte[] key = new byte[32];
//        hkdf.generateBytes(key, 0, 32);
//        return key;
//    }

File: hap_mininal8/proj/main/PairVerify.java
package main;

import java.nio.charset.StandardCharsets;
import java.security.SecureRandom;

import bouncy.*;
import crypto.*;

public class PairVerify {
    
    private static volatile SecureRandom secureRandom;

    private final AuthInfo authInfo;
    
    private byte[] hkdfKey;
    private byte[] clientPublicKey;
    private byte[] publicKey;
    private byte[] sharedSecret;
    
    private String label;
    
    public PairVerify(AuthInfo authInfo) {
        this.authInfo = authInfo;
        label = "Home Bridge";
    }


    

    public byte[] stage1(byte[] pk) throws Exception
    {
        System.out.println("Starting pair verification for " + label);
        clientPublicKey = pk; 
        publicKey = new byte[32];
        byte[] privateKey = new byte[32];
        getSecureRandom().nextBytes(privateKey);
        Curve25519.keygen(publicKey, null, privateKey);

        sharedSecret = new byte[32];
        Curve25519.curve(sharedSecret, privateKey, clientPublicKey);

        byte[] material = PairSetup.joinBytes(publicKey, authInfo.getMac().getBytes(StandardCharsets.UTF_8),
                clientPublicKey);

        byte[] proof = new EdsaSigner(authInfo.getPrivateKey()).sign(material);

        HKDFBytesGenerator hkdf = new HKDFBytesGenerator(new SHA512Digest());
        hkdf.init(new HKDFParameters(sharedSecret, "Pair-Verify-Encrypt-Salt".getBytes(StandardCharsets.UTF_8),
                "Pair-Verify-Encrypt-Info".getBytes(StandardCharsets.UTF_8)));
        hkdfKey = new byte[32];
        hkdf.generateBytes(hkdfKey, 0, 32);

        Encoder encoder = new Encoder();
        encoder.add(MessageType.USERNAME.getKey(), authInfo.getMac().getBytes(StandardCharsets.UTF_8));
        encoder.add(MessageType.SIGNATURE.getKey(), proof);
        byte[] plaintext = encoder.toByteArray();

        ChachaEncoder chacha = new ChachaEncoder(hkdfKey, "PV-Msg02".getBytes(StandardCharsets.UTF_8));
        byte[] ciphertext = chacha.encodeCiphertext(plaintext);

        encoder = new Encoder();
        encoder.add(MessageType.STATE.getKey(), (short) 2);
        encoder.add(MessageType.ENCRYPTED_DATA.getKey(), ciphertext);
        encoder.add(MessageType.PUBLIC_KEY.getKey(), publicKey);
        return encoder.toByteArray();
    }
   
    public byte[] stage2(byte[] t5) throws Exception {
        
        byte[] authData = new byte[16];
        byte[] message = new byte[t5.length - 16];
        for (int i = 0; i < 16; ++i)
        {
            authData[i] = t5[message.length + i];
        }
        for (int i = 0; i < message.length; ++i)
        {
            message[i] = t5[i];
        }
        
        ChachaDecoder chacha = new ChachaDecoder(hkdfKey, "PV-Msg03".getBytes(StandardCharsets.UTF_8));
        byte[] plaintext = chacha.decodeCiphertext(authData, message);
        
        Decoder d = new Decoder();
        d.decode(plaintext);
        byte[] clientUsername = d.getData(MessageType.USERNAME.getKey());
        byte[] clientSignature = d.getData(MessageType.SIGNATURE.getKey());
        
        byte[] material = PairSetup.joinBytes(clientPublicKey, clientUsername, publicKey);
        
        byte[] clientLtpk = authInfo.getUserPublicKey(authInfo.getMac()+new String(clientUsername, StandardCharsets.UTF_8));
        if (clientLtpk == null) {
            throw new Exception("Unknown user: "+new String(clientUsername, StandardCharsets.UTF_8));
        }
        
        Encoder encoder = new Encoder();
        if (new EdsaVerifier(clientLtpk).verify(material, clientSignature)) {
            encoder.add(MessageType.STATE.getKey(), (short) 4);
            System.out.println("Completed pair verification for " + label);
            
            authInfo.storeReadKey(createKey("Control-Write-Encryption-Key"));
            authInfo.storeWriteKey(createKey("Control-Read-Encryption-Key"));
            
            return encoder.toByteArray();
            //            return new UpgradeResponse(encoder.toByteArray(), createKey("Control-Write-Encryption-Key"),
//                    createKey("Control-Read-Encryption-Key"));
        } else {
            encoder.add(MessageType.ERROR.getKey(), (short) 4);
            throw new Exception("Invalid signature. Could not pair " + label);
        }
    }
    
    private byte[] createKey(String info) {
        HKDFBytesGenerator hkdf = new HKDFBytesGenerator(new SHA512Digest());
        hkdf.init(new HKDFParameters(sharedSecret, "Control-Salt".getBytes(StandardCharsets.UTF_8),
                info.getBytes(StandardCharsets.UTF_8)));
        byte[] key = new byte[32];
        hkdf.generateBytes(key, 0, 32);
        return key;
    }
    
    private static SecureRandom getSecureRandom() {
        if (secureRandom == null) {
            synchronized(PairVerify.class) {
                if (secureRandom == null) {
                    secureRandom = new SecureRandom();
                }
            }
        }
        return secureRandom;
    }

}
